File: src/cloud/multiplayer.ts
import { randomBytes } from 'crypto';
import { RealtimeService } from './realtime.js';
import type { RealtimeEvent } from './realtime.js';

export interface MatchConfig {
  maxPlayers: number;
  minPlayers: number;
  timeout: number;
  ranked: boolean;
  region: string;
  customData?: Record<string, unknown>;
}

export interface Match {
  id: string;
  players: string[];
  status: 'waiting' | 'ready' | 'in_progress' | 'completed' | 'cancelled';
  config: MatchConfig;
  createdAt: Date;
  startedAt?: Date;
  endedAt?: Date;
  winnerId?: string;
  serverEndpoint?: string;
}

export interface LobbyPlayer {
  userId: string;
  username: string;
  ready: boolean;
  partyId?: string;
}

export interface Lobby {
  id: string;
  name: string;
  players: LobbyPlayer[];
  config: MatchConfig;
  status: 'open' | 'in_match' | 'closed';
  hostUserId: string;
}

interface Party {
  id: string;
  members: Set<string>;
  leaderId: string;
}

class MultiplayerError extends Error {
  constructor(message: string, public readonly code: string) {
    super(message);
    this.name = 'MultiplayerError';
  }
}

export class MultiplayerService {
  private matches: Map<string, Match> = new Map();
  private lobbies: Map<string, Lobby> = new Map();
  private parties: Map<string, Party> = new Map();
  private userParty: Map<string, string> = new Map();
  private stateSync: Map<string, Set<(state: unknown) => void>> = new Map();
  private eventHandlers: Map<string, Map<string, Set<(data: unknown, sender: string) => void>>> = new Map();
  private matchmakingQueue: Map<string, { userId: string; config: MatchConfig }[]> = new Map();

  constructor(
    private readonly baseUrl: string,
    private readonly apiKey: string,
    private readonly realtime: RealtimeService
  ) {}

  async findMatch(config: MatchConfig): Promise<Match> {
    this.validateMatchConfig(config);

    const regionQueues = this.matchmakingQueue.get(config.region) ?? [];
    const entry = {
      userId: `player_${randomBytes(4).toString('hex')}`,
      config,
    };

    const compatibleIndex = regionQueues.findIndex((q) => this.areConfigsCompatible(q.config, config));
    if (compatibleIndex >= 0) {
      const matched = regionQueues.splice(compatibleIndex, 1)[0];
      const matchId = randomBytes(8).toString('hex');
      const now = new Date();
      const match: Match = {
        id: matchId,
        players: [matched.userId, entry.userId],
        status: 'ready',
        config,
        createdAt: now,
        startedAt: now,
        serverEndpoint: `${this.baseUrl}/match/${matchId}`,
      };
      this.matches.set(matchId, match);
      return { ...match };
    }

    regionQueues.push(entry);
    this.matchmakingQueue.set(config.region, regionQueues);

    const matchId = randomBytes(8).toString('hex');
    const match: Match = {
      id: matchId,
      players: [],
      status: 'waiting',
      config,
      createdAt: new Date(),
      serverEndpoint: `${this.baseUrl}/match/${matchId}`,
    };
    this.matches.set(matchId, match);
    return { ...match };
  }

  async cancelMatchmaking(): Promise<void> {
    for (const [, queue] of this.matchmakingQueue) {
      queue.length = 0;
    }
  }

  async getMatch(matchId: string): Promise<Match> {
    const match = this.matches.get(matchId);
    if (!match) {
      throw new MultiplayerError(`Match not found: ${matchId}`, 'NOT_FOUND');
    }
    return { ...match, players: [...match.players] };
  }

  async leaveMatch(matchId: string): Promise<void> {
    const match = this.matches.get(matchId);
    if (!match) {
      throw new MultiplayerError(`Match not found: ${matchId}`, 'NOT_FOUND');
    }
    match.status = 'cancelled';
    match.endedAt = new Date();
  }

  async createLobby(name: string, config: MatchConfig): Promise<Lobby> {
    this.validateMatchConfig(config);
    const lobbyId = randomBytes(8).toString('hex');
    const hostId = `host_${randomBytes(4).toString('hex')}`;
    const lobby: Lobby = {
      id: lobbyId,
      name,
      players: [
        {
          userId: hostId,
          username: `player_${hostId.slice(0, 6)}`,
          ready: false,
        },
      ],
      config,
      status: 'open',
      hostUserId: hostId,
    };
    this.lobbies.set(lobbyId, lobby);
    return this.cloneLobby(lobby);
  }

  async joinLobby(lobbyId: string): Promise<Lobby> {
    const lobby = this.lobbies.get(lobbyId);
    if (!lobby) {
      throw new MultiplayerError(`Lobby not found: ${lobbyId}`, 'NOT_FOUND');
    }
    if (lobby.status !== 'open') {
      throw new MultiplayerError('Lobby is not open', 'LOBBY_CLOSED');
    }
    if (lobby.players.length >= lobby.config.maxPlayers) {
      throw new MultiplayerError('Lobby is full', 'LOBBY_FULL');
    }

    const userId = `player_${randomBytes(4).toString('hex')}`;
    lobby.players.push({
      userId,
      username: `player_${userId.slice(0, 6)}`,
      ready: false,
    });
    return this.cloneLobby(lobby);
  }

  async leaveLobby(lobbyId: string): Promise<void> {
    const lobby = this.lobbies.get(lobbyId);
    if (!lobby) {
      throw new MultiplayerError(`Lobby not found: ${lobbyId}`, 'NOT_FOUND');
    }

    if (lobby.players.length <= 1) {
      this.lobbies.delete(lobbyId);
      return;
    }

    lobby.players.pop();
  }

  async getLobby(lobbyId: string): Promise<Lobby> {
    const lobby = this.lobbies.get(lobbyId);
    if (!lobby) {
      throw new MultiplayerError(`Lobby not found: ${lobbyId}`, 'NOT_FOUND');
    }
    return this.cloneLobby(lobby);
  }

  async listLobbies(): Promise<Lobby[]> {
    return [...this.lobbies.values()]
      .filter((l) => l.status === 'open')
      .map((l) => this.cloneLobby(l));
  }

  async setReady(lobbyId: string, ready: boolean): Promise<void> {
    const lobby = this.lobbies.get(lobbyId);
    if (!lobby) {
      throw new MultiplayerError(`Lobby not found: ${lobbyId}`, 'NOT_FOUND');
    }
    if (lobby.players.length > 0) {
      const lastPlayer = lobby.players[lobby.players.length - 1];
      lastPlayer.ready = ready;
    }
  }

  async startMatch(lobbyId: string): Promise<Match> {
    const lobby = this.lobbies.get(lobbyId);
    if (!lobby) {
      throw new MultiplayerError(`Lobby not found: ${lobbyId}`, 'NOT_FOUND');
    }
    if (lobby.players.length < lobby.config.minPlayers) {
      throw new MultiplayerError(
        `Not enough players. Need ${lobby.config.minPlayers}, have ${lobby.players.length}`,
        'NOT_ENOUGH_PLAYERS'
      );
    }
    if (lobby.players.some((p) => !p.ready)) {
      throw new MultiplayerError('Not all players are ready', 'PLAYERS_NOT_READY');
    }

    const matchId = randomBytes(8).toString('hex');
    const now = new Date();
    const match: Match = {
      id: matchId,
      players: lobby.players.map((p) => p.userId),
      status: 'in_progress',
      config: lobby.config,
      createdAt: now,
      startedAt: now,
      serverEndpoint: `${this.baseUrl}/match/${matchId}`,
    };
    this.matches.set(matchId, match);
    lobby.status = 'in_match';
    return { ...match, players: [...match.players] };
  }

  async createParty(): Promise<string> {
    const partyId = randomBytes(8).toString('hex');
    const userId = `player_${randomBytes(4).toString('hex')}`;
    const party: Party = {
      id: partyId,
      members: new Set([userId]),
      leaderId: userId,
    };
    this.parties.set(partyId, party);
    this.userParty.set(userId, partyId);
    return partyId;
  }

  async joinParty(partyId: string): Promise<void> {
    const party = this.parties.get(partyId);
    if (!party) {
      throw new MultiplayerError(`Party not found: ${partyId}`, 'NOT_FOUND');
    }
    const userId = `player_${randomBytes(4).toString('hex')}`;
    party.members.add(userId);
    this.userParty.set(userId, partyId);
  }

  async leaveParty(): Promise<void> {
    for (const [userId, partyId] of this.userParty) {
      const party = this.parties.get(partyId);
      if (party) {
        party.members.delete(userId);
        if (party.members.size === 0) {
          this.parties.delete(partyId);
        } else if (party.leaderId === userId) {
          party.leaderId = [...party.members][0];
        }
      }
      this.userParty.delete(userId);
    }
  }

  async inviteToParty(userId: string): Promise<void> {
    const partyId = this.userParty.get(userId);
    if (!partyId) {
      throw new MultiplayerError('User is not in a party', 'NOT_IN_PARTY');
    }
  }

  async syncState(matchId: string, state: unknown): Promise<void> {
    const match = this.matches.get(matchId);
    if (!match) {
      throw new MultiplayerError(`Match not found: ${matchId}`, 'NOT_FOUND');
    }
    const callbacks = this.stateSync.get(matchId);
    if (callbacks) {
      for (const cb of callbacks) {
        try {
          cb(state);
        } catch {
          continue;
        }
      }
    }
  }

  async getState(matchId: string): Promise<unknown> {
    const match = this.matches.get(matchId);
    if (!match) {
      throw new MultiplayerError(`Match not found: ${matchId}`, 'NOT_FOUND');
    }
    return match.config.customData ?? null;
  }

  onStateChange(matchId: string, callback: (state: unknown) => void): void {
    if (!this.stateSync.has(matchId)) {
      this.stateSync.set(matchId, new Set());
    }
    this.stateSync.get(matchId)!.add(callback);
  }

  async sendEvent(matchId: string, event: string, data?: unknown): Promise<void> {
    const match = this.matches.get(matchId);
    if (!match) {
      throw new MultiplayerError(`Match not found: ${matchId}`, 'NOT_FOUND');
    }
    const matchEvents = this.eventHandlers.get(matchId);
    if (matchEvents) {
      const handlers = matchEvents.get(event);
      if (handlers) {
        for (const cb of handlers) {
          try {
            cb(data, 'server');
          } catch {
            continue;
          }
        }
      }
    }
  }

  onEvent(
    matchId: string,
    eventType: string,
    callback: (data: unknown, sender: string) => void
  ): void {
    if (!this.eventHandlers.has(matchId)) {
      this.eventHandlers.set(matchId, new Map());
    }
    const matchEvents = this.eventHandlers.get(matchId)!;
    if (!matchEvents.has(eventType)) {
      matchEvents.set(eventType, new Set());
    }
    matchEvents.get(eventType)!.add(callback);
  }

  private validateMatchConfig(config: MatchConfig): void {
    if (config.maxPlayers < 2) {
      throw new MultiplayerError('maxPlayers must be at least 2', 'INVALID_CONFIG');
    }
    if (config.minPlayers < 1) {
      throw new MultiplayerError('minPlayers must be at least 1', 'INVALID_CONFIG');
    }
    if (config.minPlayers > config.maxPlayers) {
      throw new MultiplayerError('minPlayers cannot exceed maxPlayers', 'INVALID_CONFIG');
    }
    if (config.timeout <= 0) {
      throw new MultiplayerError('timeout must be positive', 'INVALID_CONFIG');
    }
  }

  private areConfigsCompatible(a: MatchConfig, b: MatchConfig): boolean {
    return (
      a.maxPlayers === b.maxPlayers &&
      a.minPlayers === b.minPlayers &&
      a.ranked === b.ranked
    );
  }

  private cloneLobby(lobby: Lobby): Lobby {
    return {
      ...lobby,
      players: lobby.players.map((p) => ({ ...p })),
      config: { ...lobby.config, customData: lobby.config.customData ? { ...lobby.config.customData } : undefined },
    };
  }
}

export { MultiplayerError };

File: zoya/cloud/multiplayer.py
import copy
import secrets
import time
from collections.abc import Callable
from dataclasses import dataclass, field
from enum import Enum
from typing import Any


class MatchStatus(str, Enum):
    WAITING = "waiting"
    READY = "ready"
    IN_PROGRESS = "in_progress"
    COMPLETED = "completed"
    CANCELLED = "cancelled"


class LobbyStatus(str, Enum):
    OPEN = "open"
    IN_MATCH = "in_match"
    CLOSED = "closed"


@dataclass
class MatchConfig:
    max_players: int = 4
    min_players: int = 2
    timeout: float = 30.0
    ranked: bool = False
    region: str = "us-east"
    custom_data: dict[str, Any] | None = None


@dataclass
class Match:
    id: str
    players: list[str] = field(default_factory=list)
    status: MatchStatus = MatchStatus.WAITING
    config: MatchConfig | None = None
    created_at: float = 0.0
    started_at: float | None = None
    ended_at: float | None = None
    winner_id: str | None = None
    server_endpoint: str | None = None


@dataclass
class LobbyPlayer:
    user_id: str
    username: str
    ready: bool = False
    party_id: str | None = None


@dataclass
class Lobby:
    id: str
    name: str
    players: list[LobbyPlayer] = field(default_factory=list)
    config: MatchConfig | None = None
    status: LobbyStatus = LobbyStatus.OPEN
    host_user_id: str | None = None


class MultiplayerError(Exception):
    def __init__(self, message: str, code: str = "MULTIPLAYER_ERROR"):
        self.code = code
        super().__init__(message)


_MOCK_USERNAMES: dict[str, str] = {
    "user_alice": "Alice",
    "user_bob": "Bob",
    "user_charlie": "Charlie",
    "user_diana": "Diana",
    "user_eve": "Eve",
    "user_frank": "Frank",
    "user_grace": "Grace",
    "user_henry": "Henry",
    "user_iris": "Iris",
    "user_jack": "Jack",
}


def _get_username(user_id: str) -> str:
    return _MOCK_USERNAMES.get(user_id, user_id)


class MultiplayerService:
    def __init__(self, base_url: str, api_key: str, realtime):
        self.base_url = base_url
        self.api_key = api_key
        self._realtime = realtime
        self._matches: dict[str, Match] = {}
        self._lobbies: dict[str, Lobby] = {}
        self._matchmaking_queue: list[str] = []
        self._parties: dict[str, set[str]] = {}
        self._player_parties: dict[str, str] = {}
        self._state: dict[str, Any] = {}
        self._state_listeners: dict[str, list[Callable]] = {}
        self._event_handlers: dict[str, list[Callable]] = {}
        self._player_lobby: dict[str, str] = {}
        self._player_match: dict[str, str] = {}
        self._matchmaking_users: set[str] = set()

    def find_match(self, config: MatchConfig) -> Match:
        if not self._matchmaking_queue:
            self._matchmaking_queue.append("user_alice")
            self._matchmaking_users.add("user_alice")

        compatible = [uid for uid in self._matchmaking_queue if uid != "user_alice"]
        if compatible:
            opponent = compatible[0]
            self._matchmaking_queue.remove(opponent)
            self._matchmaking_users.discard(opponent)
            self._matchmaking_users.discard("user_alice")
            self._matchmaking_queue.remove("user_alice")

            match = Match(
                id=secrets.token_hex(8),
                players=["user_alice", opponent],
                status=MatchStatus.IN_PROGRESS,
                config=config,
                created_at=time.time(),
                started_at=time.time(),
                server_endpoint=f"{self.base_url}/match/{secrets.token_hex(4)}",
            )
            self._matches[match.id] = match
            for uid in match.players:
                self._player_match[uid] = match.id
            return match

        self._matchmaking_queue.append("user_alice")
        self._matchmaking_users.add("user_alice")

        return Match(
            id="pending",
            players=["user_alice"],
            status=MatchStatus.WAITING,
            config=config,
            created_at=time.time(),
        )

    def cancel_matchmaking(self) -> None:
        self._matchmaking_queue[:] = [
            u for u in self._matchmaking_queue if u != "user_alice"
        ]
        self._matchmaking_users.discard("user_alice")

    def get_match(self, match_id: str) -> Match:
        match = self._matches.get(match_id)
        if match is None:
            raise MultiplayerError(
                f"Match '{match_id}' not found", code="MATCH_NOT_FOUND"
            )
        return copy.deepcopy(match)

    def leave_match(self, match_id: str) -> None:
        match = self._matches.get(match_id)
        if match is None:
            raise MultiplayerError(
                f"Match '{match_id}' not found", code="MATCH_NOT_FOUND"
            )
        match.status = MatchStatus.CANCELLED
        match.ended_at = time.time()
        for uid in match.players:
            self._player_match.pop(uid, None)

    def create_lobby(self, name: str, config: MatchConfig) -> Lobby:
        lobby_id = secrets.token_hex(8)
        lobby = Lobby(
            id=lobby_id,
            name=name,
            config=config,
            host_user_id="user_alice",
        )
        self._lobbies[lobby_id] = lobby
        return lobby

    def join_lobby(self, lobby_id: str) -> Lobby:
        lobby = self._lobbies.get(lobby_id)
        if lobby is None:
            raise MultiplayerError(
                f"Lobby '{lobby_id}' not found", code="LOBBY_NOT_FOUND"
            )
        if lobby.status != LobbyStatus.OPEN:
            raise MultiplayerError("Lobby is not open", code="LOBBY_CLOSED")

        party_id = self._player_parties.get("user_alice")
        player = LobbyPlayer(
            user_id="user_alice",
            username=_get_username("user_alice"),
            party_id=party_id,
        )
        lobby.players.append(player)
        self._player_lobby["user_alice"] = lobby_id
        return lobby

    def leave_lobby(self, lobby_id: str) -> None:
        lobby = self._lobbies.get(lobby_id)
        if lobby is None:
            raise MultiplayerError(
                f"Lobby '{lobby_id}' not found", code="LOBBY_NOT_FOUND"
            )
        lobby.players = [p for p in lobby.players if p.user_id != "user_alice"]
        self._player_lobby.pop("user_alice", None)

    def get_lobby(self, lobby_id: str) -> Lobby:
        lobby = self._lobbies.get(lobby_id)
        if lobby is None:
            raise MultiplayerError(
                f"Lobby '{lobby_id}' not found", code="LOBBY_NOT_FOUND"
            )
        return copy.deepcopy(lobby)

    def list_lobbies(self) -> list[Lobby]:
        return [copy.deepcopy(l) for l in self._lobbies.values()]

    def set_ready(self, lobby_id: str, ready: bool) -> None:
        lobby = self._lobbies.get(lobby_id)
        if lobby is None:
            raise MultiplayerError(
                f"Lobby '{lobby_id}' not found", code="LOBBY_NOT_FOUND"
            )
        for player in lobby.players:
            if player.user_id == "user_alice":
                player.ready = ready
                return
        raise MultiplayerError("User not in lobby", code="USER_NOT_IN_LOBBY")

    def start_match(self, lobby_id: str) -> Match:
        lobby = self._lobbies.get(lobby_id)
        if lobby is None:
            raise MultiplayerError(
                f"Lobby '{lobby_id}' not found", code="LOBBY_NOT_FOUND"
            )
        if lobby.host_user_id != "user_alice":
            raise MultiplayerError("Only the host can start a match", code="NOT_HOST")

        ready_count = sum(1 for p in lobby.players if p.ready)
        min_players = lobby.config.min_players if lobby.config else 2
        if ready_count < min_players:
            raise MultiplayerError(
                f"Not enough ready players ({ready_count}/{min_players})",
                code="NOT_ENOUGH_PLAYERS",
            )

        match = Match(
            id=secrets.token_hex(8),
            players=[p.user_id for p in lobby.players],
            status=MatchStatus.IN_PROGRESS,
            config=lobby.config,
            created_at=time.time(),
            started_at=time.time(),
            server_endpoint=f"{self.base_url}/match/{secrets.token_hex(4)}",
        )
        self._matches[match.id] = match
        lobby.status = LobbyStatus.IN_MATCH
        return match

    def create_party(self) -> str:
        party_id = secrets.token_hex(8)
        self._parties[party_id] = {"user_alice"}
        self._player_parties["user_alice"] = party_id
        return party_id

    def join_party(self, party_id: str) -> None:
        if party_id not in self._parties:
            self._parties[party_id] = set()
        self._parties[party_id].add("user_alice")
        self._player_parties["user_alice"] = party_id

    def leave_party(self) -> None:
        party_id = self._player_parties.pop("user_alice", None)
        if party_id and party_id in self._parties:
            self._parties[party_id].discard("user_alice")
            if not self._parties[party_id]:
                del self._parties[party_id]

    def invite_to_party(self, user_id: str) -> None:
        party_id = self._player_parties.get("user_alice")
        if party_id is None:
            raise MultiplayerError("User is not in a party", code="NO_PARTY")

    def sync_state(self, match_id: str, state: Any) -> None:
        self._state[match_id] = state
        for callback in self._state_listeners.get(match_id, []):
            callback(state)

    def get_state(self, match_id: str) -> Any:
        return copy.deepcopy(self._state.get(match_id))

    def on_state_change(self, match_id: str, callback: Callable) -> None:
        if match_id not in self._state_listeners:
            self._state_listeners[match_id] = []
        self._state_listeners[match_id].append(callback)

    def send_event(self, match_id: str, event: str, data: Any = None) -> None:
        for callback in self._event_handlers.get(f"{match_id}:{event}", []):
            callback(data)
        for callback in self._event_handlers.get(f"{match_id}:*", []):
            callback({"event": event, "data": data})

    def on_event(self, match_id: str, event_type: str, callback: Callable) -> None:
        key = f"{match_id}:{event_type}"
        if key not in self._event_handlers:
            self._event_handlers[key] = []
        self._event_handlers[key].append(callback)
